lib/Sema/CSWalker.cpp

```diff
@@ -218,35 +218,199 @@ class LocalCSWalker : public glu::ast::ASTWalker<LocalCSWalker, void> {
+    /// FunctionDecl.
+    /// @returns True if any overload constraints were added; false if none
+    /// matched.
+    bool resolveOverloadSet(
+        llvm::StringRef name, glu::ast::ExprBase *anchor, bool isExplicit,
+        llvm::function_ref<Constraint *(glu::ast::FunctionDecl *)>
+            constraintBuilder
+    )
+    {
+        auto *item = _cs.getScopeTable()->lookupItem(name);
+        llvm::ArrayRef<glu::ast::DeclBase *> decls
+            = item ? item->decls : llvm::ArrayRef<glu::ast::DeclBase *>();
+
+        llvm::SmallVector<Constraint *, 4> constraints;
+
+        for (auto *decl : decls) {
+            if (auto *fnDecl = llvm::dyn_cast<glu::ast::FunctionDecl>(decl)) {
+                if (auto *constraint = constraintBuilder(fnDecl)) {
+                    constraints.push_back(constraint);
+                }
+            }
         }
 
         if (constraints.empty()) {
-            llvm::Twine errorMsg = llvm::Twine("use of undeclared operator '")
-                + node->getOperator().getLexeme() + llvm::Twine("'");
+            return false;
+        }
+
+        auto *disjunction = Constraint::createDisjunction(
+            _cs.getAllocator(), constraints, anchor, isExplicit
+        );
+        _cs.addConstraint(disjunction);
+        return true;
+    }
+
+    /// @brief Processes variable declarations that might contain function
+    /// references
+    void processVariableDeclarations(
+        glu::ast::CallExpr *node, llvm::ArrayRef<glu::ast::DeclBase *> decls
+    )
+    {
+        for (auto decl : decls) {
+            if (auto varLetDecl = llvm::dyn_cast<glu::ast::VarLetDecl>(decl)) {
+                auto declType = varLetDecl->getType();
+                if (!declType)
+                    continue;
+
+                glu::types::FunctionTy *functionTy
+                    = extractFunctionType(declType);
+
+                if (functionTy) {
+                    _cs.addConstraint(
+                        Constraint::createEqual(
+                            _cs.getAllocator(), node->getType(),
+                            functionTy->getReturnType(), node
+                        )
+                    );
+                    constrainArguments(node, functionTy);
+                }
+            }
+        }
+    }
 
-            _diagManager.error(node->getLocation(), std::move(errorMsg));
+    /// @brief Handles function calls through function pointers
+    void handlePointerCall(glu::ast::CallExpr *node)
+    {
+        auto calleeType = node->getCallee()->getType();
+        if (!calleeType)
+            return;
+
+        glu::types::FunctionTy *functionTy = extractFunctionType(calleeType);
+
+        if (functionTy) {
+            _cs.addConstraint(
+                Constraint::createEqual(
+                    _cs.getAllocator(), node->getType(),
+                    functionTy->getReturnType(), node
+                )
+            );
```

Comment:
the constraint should be about conversion from a function type to another function type